osbuild/util/pe32p.py

```diff
@@ -0,0 +1,203 @@
+file positions, but it requires that the functions are called in the given
+order, i.e. `read_coff_header`, `read_optional_haeder` then `read_sections`.
+
+The format definition follows https://learn.microsoft.com/en-us/windows/win32/debug/pe-format
+"""
+
+import enum
+import io
+import os
+import struct
+import sys
+from collections import namedtuple
+from typing import BinaryIO, Iterator, List, Optional, Union
+
+PathLike = Union[str, bytes, os.PathLike]
+
+CoffFormat = "4sHHIIIHH"
+CoffHeader = namedtuple(
+    "CoffHeader",
+    [
+        "Signature",
+        "Machine",
+        "NumberOfSections",
+        "TimeDateStamp",
+        "PointerToSymbolTable",
+        "NumberOfSymbols",
+        "SizeOfOptionalHeader",
+        "Characteristics",
+    ]
+)
+
+
+SectionFormat = "8sIIIIIIHHI"
+SectionHeader = namedtuple(
+    "SectionHeader",
+    [
+        "Name",
+        "VirtualSize",
+        "VirtualAddress",
+        "SizeOfRawData",
+        "PointerToRawData",
+        "PointerToRelocations",
+        "PointerToLinenumbers",
+        "NumberOfRelocations",
+        "NumberOfLinenumbers",
+        "Characteristics",
+    ]
+)
+
+
+class SectionFlags(enum.Flag):
+    ALIGN_1BYTES = 0x00100000
+    ALIGN_2BYTES = 0x00200000
+    ALIGN_4BYTES = 0x00300000
+    ALIGN_8BYTES = 0x00400000
+    ALIGN_16BYTES = 0x00500000
+    ALIGN_32BYTES = 0x00600000
+    ALIGN_64BYTES = 0x00700000
+    ALIGN_128BYTES = 0x00800000
+    ALIGN_256BYTES = 0x00900000
+    ALIGN_512BYTES = 0x00A00000
+    ALIGN_1024BYTES = 0x00B00000
+    ALIGN_2048BYTES = 0x00C00000
+    ALIGN_4096BYTES = 0x00D00000
+    ALIGN_8192BYTES = 0x00E00000
+    ALIGN_MASK = 0x00F00000
+    ALIGN_DEFAULT = ALIGN_16BYTES
+
+
+OptionalFormat = "HBBIIIIIQIIHHHHHHIIIIHHQQQQII"
+OptionalHeader = namedtuple(
+    "OptionalHeader",
+    [
+        # Standard fields
+        "Magic",
+        "MajorLinkerVersion",
+        "MinorLinkerVersion",
+        "SizeOfCode",
+        "SizeOfInitializedData",
+        "SizeOfUninitializedData",
+        "AddressOfEntryPoint",
+        "BaseOfCode",
+        # Windows-Specific fields (PE32+)
+        "ImageBase",
+        "SectionAlignment",
+        "FileAlignment",
+        "MajorOperatingSystemVersion",
+        "MinorOperatingSystemVersion",
+        "MajorImageVersion",
+        "MinorImageVersion",
+        "MajorSubsystemVersion",
+        "MinorSubsystemVersion",
+        "Reserved1",
+        "SizeOfImage",
+        "SizeOfHeaders",
+        "CheckSum",
+        "Subsystem",
+        "DllCharacteristics",
+        "SizeOfStackReserve",
+        "SizeOfStackCommit",
+        "SizeOfHeapReserve",
+        "SizeOfHeapCommit",
+        "LoaderFlags",
+        "NumberOfRvaAndSizes",
+    ]
+)
+
+
+def _seek_to_coff(f: BinaryIO, check: bool = False) -> None:
+    if check:
+        f.seek(0, io.SEEK_SET)
+        magic = f.read(2)
+        assert magic == b"MZ", "Not a PE32+ file (missing DOS stub)"
```

Comment:
Strictly speaking, this is not correct. You can place non-MZ stubs in front of a PE32+ file just fine. You just need to make sure the PE-offset is stored at `0x3c`. Note that even in MZ-stubs the offset `0x3c` is not special in any way, but just happens to be neatly the last 4 bytes of the first 64-byte block, and thus commonly used for padding before storing the relocations.

Anyway, I would suggest dropping this and just checking that the PE-signature is present at the indicated location.

Reply:
I removed the MZ check.